src/main/java/bio/terra/service/filedata/azure/tables/TableFileDao.java

```diff
@@ -0,0 +1,157 @@
+    tableServiceClient.createTableIfNotExists(TABLE_NAME);
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = FireStoreFile.toTableEntity(PARTITION_KEY, newFile);
+    logger.info("creating file metadata for fileId: " + newFile.getFileId());
+    tableClient.createEntity(entity);
+  }
+
+  public boolean deleteFileMetadata(TableServiceClient tableServiceClient, String fileId) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = tableClient.getEntity(PARTITION_KEY, fileId);
+    if (entity == null) {
+      return false;
+    }
+    logger.info("deleting file metadata for fileId " + fileId);
+    tableClient.deleteEntity(PARTITION_KEY, fileId);
+    return true;
+  }
+
+  public FireStoreFile retrieveFileMetadata(TableServiceClient tableServiceClient, String fileId) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity = tableClient.getEntity(PARTITION_KEY, fileId);
+    return FireStoreFile.fromTableEntity(entity);
+  }
+
+  /**
+   * Retrieve metadata from a list of directory entries.
+   *
+   * @param tableServiceClient An Azure table service client
+   * @param directoryEntries List of objects to retrieve metadata for
+   * @return A list of metadata object for the specified files. Note: the order of the list matches
+   *     with the order of the input list objects
+   */
+  List<FireStoreFile> batchRetrieveFileMetadata(
+      TableServiceClient tableServiceClient, List<FireStoreDirectoryEntry> directoryEntries) {
+    return directoryEntries.stream()
+        .map(
+            f ->
+                Optional.ofNullable(retrieveFileMetadata(tableServiceClient, f.getFileId()))
+                    .orElseThrow(
+                        () ->
+                            new FileSystemCorruptException(
+                                "Directory entry refers to non-existent file")))
+        .collect(Collectors.toList());
+  }
+
+  <V> void scanTableObjects(TableClient tableClient, ApiFutureGenerator<V, TableEntity> generator) {
+    ListEntitiesOptions options = new ListEntitiesOptions();
+    PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
+    entities.stream()
+        .forEach(
+            entity -> {
+              try {
+                retryOperation(entity, generator);
+              } catch (InterruptedException e) {
+                throw new FileSystemExecutionException("operation failed", e);
+              }
+            });
+  }
+
+  <T, V> void retryOperation(V input, ApiFutureGenerator<T, V> generator)
+      throws InterruptedException {
+    int retry = 0;
+    while (true) {
+      try {
+        generator.accept(input).get();
+        break;
+      } catch (ExecutionException ex) {
+        final long retryWait = (long) (SLEEP_BASE_SECONDS * Math.pow(2.5, retry));
```

Comment:
Is there any way to reuse this?  I guess it's got some azure specific logic in there.  It might make sense to put a ceiling on the amount between retries (this gets to a point quickly where it waits an hour between retries.  Maybe max it out at a couple of minutes between retries?